subcerts.go

```diff
@@ -0,0 +1,515 @@
+	var sig []byte
+	switch sk := cert.PrivateKey.(type) {
+	case *ecdsa.PrivateKey:
+		opts := crypto.SignerOpts(hash)
+		sig, err = sk.Sign(rand.Reader, in, opts)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, fmt.Errorf("unsupported delgation key type: %T", sk)
+	}
+
+	return &DelegatedCredential{
+		ValidTime:       validTime,
+		PublicKey:       pk,
+		PrivateKey:      sk,
+		publicKeyScheme: scheme,
+		Scheme:          delegatorScheme,
+		Signature:       sig,
+	}, nil
+}
+
+// IsExpired returns true if the credential has expired. The end of the validity
+// interval is defined as the delegator certificate's notBefore field (`start`)
+// plus ValidTime seconds. This function simply checks that the current time
+// (`now`) is before the end of the valdity interval.
+func (dc *DelegatedCredential) IsExpired(start, now time.Time) bool {
+	end := start.Add(dc.ValidTime)
+	return !now.Before(end)
+}
+
+// InvalidTTL returns true if the credential's validity period is longer than the
+// maximum permitted. This is defined by the certificate's notBefore field
+// (`start`) plus the ValidTime, minus the current time (`now`).
+func (dc *DelegatedCredential) InvalidTTL(start, now time.Time) bool {
+	return dc.ValidTime > (now.Sub(start) + dcMaxTTL).Round(time.Second)
+}
+
+// Validate checks that that the signature is valid, that the credential hasn't
+// expired, and that the TTL is valid. It also checks that certificate can be
+// used for delegation.
+func (dc *DelegatedCredential) Validate(cert *x509.Certificate, vers uint16, now time.Time) (bool, error) {
+	// Check that the cert can delegate.
+	if !canDelegate(cert) {
+		return false, errNoDelegationUsage
+	}
+
+	if dc.IsExpired(cert.NotBefore, now) {
+		return false, errors.New("credential has expired")
+	}
+
+	if dc.InvalidTTL(cert.NotBefore, now) {
+		return false, errors.New("credential TTL is invalid")
+	}
+
+	// Prepare the credential for verification.
+	hash := getHash(dc.Scheme)
+	cred := &credential{dc.ValidTime, dc.PublicKey, dc.publicKeyScheme}
+	in, err := prepareDelegation(hash, cred, cert.Raw, dc.Scheme, vers)
+	if err != nil {
+		return false, err
+	}
+
+	// TODO(any) This code overlaps signficantly with verifyHandshakeSignature()
+	// in ../auth.go. This should be refactored.
+	switch dc.Scheme {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512:
+		pk, ok := cert.PublicKey.(*ecdsa.PublicKey)
+		if !ok {
+			return false, errors.New("expected ECDSA public key")
+		}
+		sig := new(ecdsaSignature)
+		if _, err = asn1.Unmarshal(dc.Signature, sig); err != nil {
+			return false, err
+		}
+		return ecdsa.Verify(pk, in, sig.R, sig.S), nil
+
+	default:
+		return false, fmt.Errorf(
+			"unsupported signature scheme: 0x%04x", dc.Scheme)
+	}
+}
+
+// Marshal encodes a DelegatedCredential structure per the spec. It also sets
+// dc.Raw to the output as a side effect.
+func (dc *DelegatedCredential) Marshal() ([]byte, error) {
```

Comment:
crypto/tls doesn't provide any other marshalers / unmarshalers. Why does it make sense to have these exported?

Reply:
Why does somebody outside of crypto/tls need to be able to compute the TLS wire-format DC?

Reply:
We need to be able to use these APIs from Keyless, for example.